Generalize constants so they also support words

paracetamolo opened the pull request:
Fixes #1856.

Because in the past all types supported by `push` would fit a Felt, there are some invariants in the current code that are not working well now that it also supports words.

In particular given that at parsing time we can't always determined the type of a `ConstantName`, we need a way to encode a push in the ast that works for any supported type. After semantic analysis, once all constants are propagated, the generic push will be resolved to the right one (e.g. push of a single felt or push of a word, which is 4 push of a felt).

Right now this generic instruction is `Push(ImmFelt)` which is limited to a single Felt. 

This PR introduces an enum (temporarily called `All`) for all possible types supported by push and changes the generic push instruction to `Push(Immediate<All>)`.
During assembly the right sequence of pushes is emitted by pattern matching on the cases of `All`.

There is some overlap between `HexEncodedValue` and `All`, so they might be merged.

bitwalker commented on assembly/src/ast/constants.rs:
```diff
@@ -242,6 +244,10 @@ impl crate::prettier::PrettyPrint for ConstantExpr {
 
         match self {
             Self::Literal(literal) => display(literal),
+            Self::Word(spanned) => {
```
I'm not sure that this will display words incorrectly. I can't recall if the default implementation of `Display` for `Felt` is in decimal or hex format, but either way I suspect this will produce output that is not valid input (i.e. cannot be round-tripped).

* If it's rendered in decimal form, MASM syntax expects the felts to be dot-separated, i.e. `A.B.C.D`.
* If it's rendered in hex form, there are two potential problems: 1.) I _think_ the defaults prefix the output with `0x`, so this would produce a string like `0x..0x..0x..0x..`, but I'm not 100% on that. 2.) MASM syntax expects words in hex-encoded form to be big-endian, we should ensure that when `Felt` is rendered as hex, that the bytes are in big-endian order. Again, that may already be the case, but I can't recall for sure.

Anyway, I'd make sure that when the AST is pretty-printed, that the output is still valid MASM syntax.

paracetamolo replied:
You are right that there was a problem but it was somewhere else, `Token::HexEncodedValue` (or now `IntValue`). Now they should all be consistent.

plafer replied:
Why `IntValue`? This name doesn't seem to work for the `Felt` and `Word` variants (unless I'm reading it wrong)

paracetamolo replied:
We use `HexEncodedValue` for anything that can fit a felt or a word, regardless of which representation it comes from, hex or not. I don't mind long explicit names so `FeltOrWord` is ok for me if you prefer.

plafer replied:
So it's used for all "values that were typed" right? Since this is in the context of the lexer, maybe just `Value` is good enough?

plafer replied:
I can't say that the first thing that comes to mind when thinking of a `Word` is "encoding a large integer" (even though you can theoretically push this to the extreme and say that any string of bits is an integer).

But yes naming is...hard..., and if both of you like it then I'm good with it

bitwalker commented on assembly/src/ast/instruction/mod.rs:
```diff
@@ -204,7 +205,7 @@ pub enum Instruction {
 
     // ----- input / output operations -----------------------------------------------------------
-    Push(ImmFelt),
+    Push(Immediate<HexEncodedValue>),
```
I think we should use this opportunity to rename `HexEncodedValue` to `IntValue`, since it is used now in more places than just the handling of hex-encoded integers.

bitwalker reviewed (approved):
Looks good! Just a couple of small suggestions, but overall 👍

bitwalker commented on assembly/src/ast/tests.rs:
```diff
@@ -1288,3 +1288,34 @@ end
+#[test]
+fn test_words_roundtrip_formatting() {
+    let source = "\
+const.A=0x0200000000000000020000000000000002000000000000000200000000000000
```
> I find this really hard to parse. Isn't there a more "array-like" syntax that would work, even if it had to be `const.A=2.2.2.2`?. Also, we expected `push.A` to be equivalent to `push.2.2.2.2`

Currently, the only way to express words as constants in MASM is via big-endian hex-encoded strings that must be of exactly the right number of bytes, this PR just builds off the existing support for parsing words.

The `push.A.B.C.D` syntax isn't a word-oriented thing, but rather syntax sugar for pushing multiple elements in the given order. We _could_ add syntax sugar to allow defining word-like constants that way, but it would complicate the grammar a bit (we couldn't just build off of the existing non-terminal for hex-encoded values).

> 1. How does `0x0200000000000000` translate to the first `2`?

This does indeed appear wrong to me, as this should be big-endian encoded (and thus the most-significant byte should appear furthest left, but here it is the least-significant byte that comes first).

However, I think this _is_ actually big-endian encoded, but the issue is rather how the value is being zero-padded out to 16 bytes. The padding should come first, but it seems to be trailing. I might be wrong about that, but I have no idea why we would ever be deliberately printing little-endian hex, it is highly unintuitive.

> 2. The test should use different values to test for ordering as well, e.g. would map to `push.2.3.4.5`

Agreed, we definitely want to verify the ordering is what we expect.

paracetamolo commented:
Sorry I needed to do another significant refactor. The problem is that there was an ambiguity in the grammar between words as dot-delimited values e.g. `2.2.2.2` and pushing multiple elements e.g. `push 2.2.2.2 => push.2 push.2 push.2 push.2`.
So because of this and of the received feedback, I introduced an array-like syntax for words as `[f0,f1,f2,f3]`. Now constants can be correctly defined as words, which was not the case before.
Words can also be defined as hex strings of 64 characters.
I also introduced the type `Word` in the parser in order to define the pretty printers which were a bit scattered and hard to keep consistent.

plafer commented on assembly/src/ast/tests.rs:
```diff
@@ -1288,3 +1289,45 @@ end
+    let formatted = module.to_string();
+    let expected = "\
+begin
+    push.[2,2,2,2]
```
nit: Would it not be more consistent to render this as a series of `push.2`? Basically semantically `push.2.2.2.2` is equivalent to `push.A` or `push.B`, but the first is "destructured" while the second isn't.

paracetamolo commented on assembly/src/ast/tests.rs:
```diff
@@ -1288,3 +1289,45 @@ end
+#[test]
+fn test_words_roundtrip_formatting() {
+    let source = "\
+const.A=0x0200000000000000020000000000000002000000000000000200000000000000
+const.B=[2,2,2,2]
+begin
+    push.0x0200000000000000020000000000000002000000000000000200000000000000.2
+    push.A.2
+    push.B.2
+    push.2.2.2.2
+end
```
done

paracetamolo commented on assembly/src/ast/visit.rs:
```diff
@@ -621,6 +625,13 @@ where
     ControlFlow::Continue(())
 }
 
+pub fn visit_immediate_hex<V, T>(_visitor: &mut V, _imm: &Immediate<IntValue>) -> ControlFlow<T>
```
done

paracetamolo commented on assembly/src/tests.rs:
```diff
@@ -2614,6 +2614,26 @@ end";
+        "\
+const.A=0x0200000000000000020000000000000002000000000000000200000000000000
+begin
+    push.A
```
I added a test. Parsing goes through and then during semantic analysis we realize it's the wrong type and raise an error.

paracetamolo commented on assembly/src/assembler/instruction/mod.rs:
```diff
@@ -336,12 +336,18 @@ impl Assembler {
 
             // ----- input / output instructions --------------------------------------------------
-            Instruction::Push(imm) => env_ops::push_one(imm.expect_value(), block_builder),
+            Instruction::Push(imm) => match (*imm).expect_value() {
+                IntValue::U8(v) => env_ops::push_one(v, block_builder),
+                IntValue::U16(v) => env_ops::push_one(v, block_builder),
+                IntValue::U32(v) => env_ops::push_one(v, block_builder),
+                IntValue::Felt(v) => env_ops::push_one(v, block_builder),
+                IntValue::Word(v) => env_ops::push_many(&v.0, block_builder),
+            },
             Instruction::PushU8(imm) => env_ops::push_one(*imm, block_builder),
             Instruction::PushU16(imm) => env_ops::push_one(*imm, block_builder),
             Instruction::PushU32(imm) => env_ops::push_one(*imm, block_builder),
             Instruction::PushFelt(imm) => env_ops::push_one(*imm, block_builder),
```
I think it's for historical reasons. I can look into simplifying it.

bitwalker replied:
It predates a lot of the parser/AST changes. I don't know that there is any benefit in having any of them other than `Push` at this point, as they could all be represented by `Push` now that we have word constants.

bobbinth replied:
@paracetamolo - let's make a quick follow-up PR to clean this up.

bobbinth commented on assembly/src/parser/token.rs:
```diff
@@ -32,13 +32,68 @@ impl core::ops::Deref for DocumentationType {
     }
 }
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct Word(pub [Felt; 4]);
```
Right, but in the next version of `miden-crypto` [Word](https://github.com/0xMiden/crypto/blob/next/miden-crypto/src/word.rs#L32) is a struct (not just a type alias).

bitwalker replied:
I can take a look at it and make any necessary updates, but IIRC it doesn't have much (if anything) in the way of dependencies.

paracetamolo commented:
> push.A.B # is this valid?

Yes this is valid, I added it to the test.

bitwalker reviewed (approved):
Looks great, nice work! Just one small change I'd like to make before merging, but I'm marking this approved to avoid holding things up once you're ready.
